MoltenVK/MoltenVK/GPUObjects/MVKDevice.mm

```diff
@@ -5324,164 +5324,65 @@ static uint32_t mvkGetEntryProperty(io_registry_entry_t entry, CFStringRef prope
-				enableFeatures(requestedFeatures,
-							   &_enabledDescriptorIndexingFeatures.shaderInputAttachmentArrayDynamicIndexing,
-							   &requestedFeatures->shaderInputAttachmentArrayDynamicIndexing,
-							   &pdDescriptorIndexingFeatures.shaderInputAttachmentArrayDynamicIndexing, 20);
-				enableFeatures(requestedFeatures,
-							   &_enabledVulkan12FeaturesNoExt.samplerFilterMinmax,
-							   &requestedFeatures->samplerFilterMinmax,
-							   &pdvulkan12FeaturesNoExt.samplerFilterMinmax, 1);
-				enableFeatures(requestedFeatures,
-							   &_enabledScalarBlockLayoutFeatures.scalarBlockLayout,
-							   &requestedFeatures->scalarBlockLayout,
-							   &pdScalarBlockLayoutFeatures.scalarBlockLayout, 1);
-				enableFeatures(requestedFeatures,
-							   &_enabledImagelessFramebufferFeatures.imagelessFramebuffer,
-							   &requestedFeatures->imagelessFramebuffer,
-							   &pdImagelessFramebufferFeatures.imagelessFramebuffer, 1);
-				enableFeatures(requestedFeatures,
-							   &_enabledUniformBufferStandardLayoutFeatures.uniformBufferStandardLayout,
-							   &requestedFeatures->uniformBufferStandardLayout,
-							   &pdUniformBufferStandardLayoutFeatures.uniformBufferStandardLayout, 1);
-				enableFeatures(requestedFeatures,
-							   &_enabledShaderSubgroupExtendedTypesFeatures.shaderSubgroupExtendedTypes,
-							   &requestedFeatures->shaderSubgroupExtendedTypes,
-							   &pdShaderSubgroupExtendedTypesFeatures.shaderSubgroupExtendedTypes, 1);
-				enableFeatures(requestedFeatures,
-							   &_enabledSeparateDepthStencilLayoutsFeatures.separateDepthStencilLayouts,
-							   &requestedFeatures->separateDepthStencilLayouts,
-							   &pdSeparateDepthStencilLayoutsFeatures.separateDepthStencilLayouts, 1);
-				enableFeatures(requestedFeatures,
-							   &_enabledHostQueryResetFeatures.hostQueryReset,
-							   &requestedFeatures->hostQueryReset,
-							   &pdHostQueryResetFeatures.hostQueryReset, 1);
-				enableFeatures(requestedFeatures,
-							   &_enabledTimelineSemaphoreFeatures.timelineSemaphore,
-							   &requestedFeatures->timelineSemaphore,
-							   &pdTimelineSemaphoreFeatures.timelineSemaphore, 1);
-				enableFeatures(requestedFeatures,
-							   &_enabledBufferDeviceAddressFeatures.bufferDeviceAddress,
-							   &requestedFeatures->bufferDeviceAddress,
-							   &pdBufferDeviceAddressFeatures.bufferDeviceAddress, 3);
-				enableFeatures(requestedFeatures,
-							   &_enabledVulkanMemoryModelFeatures.vulkanMemoryModel,
-							   &requestedFeatures->vulkanMemoryModel,
-							   &pdVulkanMemoryModelFeatures.vulkanMemoryModel, 3);
-				enableFeatures(requestedFeatures,
-							   &_enabledVulkan12FeaturesNoExt.shaderOutputViewportIndex,
-							   &requestedFeatures->shaderOutputViewportIndex,
-							   &pdvulkan12FeaturesNoExt.shaderOutputViewportIndex, 3);
+				enablePromotedFeatures(Vulkan12NoExt, samplerMirrorClampToEdge, 2);
+				enablePromotedFeatures(8BitStorage, storageBuffer8BitAccess, 3);
```

Comment:
I'm amazed this works, given that the lexical rules for C++ say that an identifier can't start with a numeral. I guess the preprocessor is OK with that as long as it gets pasted into an identifier that _is_ legal.

Reply:
Yeah. I had a brief "oh damn this won't work" moment when I got to `8BitStorage`, but then I remembered that we had used numeric starts elsewhere already. 

It actually serves as a reminder that the preprocessor really is just a text-substitution mechanism.